src/onnx/parse_if.cpp

```diff
@@ -65,13 +67,108 @@ struct parse_if : op_parser<parse_if>
 
         auto then_out_shapes = then_mdl->get_output_shapes();
         auto else_out_shapes = else_mdl->get_output_shapes();
-        if(not std::equal(then_out_shapes.begin(),
-                          then_out_shapes.end(),
-                          else_out_shapes.begin(),
-                          else_out_shapes.end()))
-        {
+
+        auto throw_shapes = [&]() {
             MIGRAPHX_THROW("PARSE_IF: " + info.name +
-                           " then and else sub_grahps must have same output shapes!");
+                           " then and else sub_graphs must have compatible shapes ");
+        };
+
+        if(then_out_shapes.size() != else_out_shapes.size())
+        {
+            throw_shapes();
+        }
+
+        // Add checks for each output shape
+        for(int i = 0; i < then_out_shapes.size(); i++)
+        {
+            // Must have the same type for both if/else blocks by onnx spec
```

Comment:
Index at the top of the loop and assign everything to a variable because `.at` does a bounds check everytime you access it:

```cpp
const auto& then_out_shape = then_out_shapes.at(i);
const auto& else_out_shape = else_out_shapes.at(i);
```

Reply:
Ah! Right. Good catch. Fixing this

Reply:
Done